### PR TITLE
CQL time types representation

### PR DESCRIPTION
**Pull request by asledz**
## Description

PR changes CQL time types in a more user-friendly way (before they were represented by raw integer).
Now, time is represented by standard Rust from https://doc.rust-lang.org/std/time/index.html.
Date is represented by chrono date.

(Timestamp, because of it specification, stays as it is for now - it is represented by a number by definition)

### Fixes
Fixes issue: #182

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla/src/frame/cql_types_test.rs [hunk 1]
```diff
@@ -25,15 +26,15 @@ async fn test_cql_types() {
...
+        .query("INSERT INTO ks.days (day, id) VALUES (?, 1)", (date,))
         .await
         .unwrap();
 
     if let Some(rows) = session.query("SELECT day FROM ks.days", &[]).await.unwrap() {
-        for row in rows.into_typed::<(u32,)>() {
-            let day: u32 = row.unwrap().0;
+        for row in rows.into_typed::<(NaiveDate,)>() {
+            let day: NaiveDate = row.unwrap().0;
             println!("day: {}", day);
```
Comment (psarna) [inline scylla/src/frame/cql_types_test.rs:38]: Looks good, but please also assert that the value is what's expected (1.1.2020).

File: scylla/src/frame/response/result.rs [hunk 2]
```diff
@@ -475,7 +478,11 @@ fn deser_cql_value(typ: &ColumnType, buf: &mut &[u8]) -> StdResult<CQLValue, Par
                     buf.len()
                 )));
             }
-            CQLValue::Date(buf.read_u32::<BigEndian>()?)
+            let timestamp = buf.read_u32::<BigEndian>()?;
+            //  An unsigned integer representing days with epoch centered at 2^31. (unix epoch January 1st, 1970).
+            let time =
+                NaiveDate::from_ymd(-5877641, 6, 23) + DurationDateTime::days(timestamp.into());
```
Comment (psarna) [inline scylla/src/frame/response/result.rs:484]: This magic value is used in two places (here and value.rs) - please put it as a const somewhere (with a descriptive name), along with the comment explaining why it's needed.

**Review by psarna — COMMENTED** [review]
Looks good, I had minor comments.

**Comment by cvybhu** [comment]
@psarna
I made some changes and added some tests
I feel it's not the prettiest but passable

### Date
* Changed `CQLValue` representation to `u32` to allow receiving full range of dates allowed by the spec.
`NaiveDate` can still be received using `into_typed` and sent as a bound value
* For cases in which the users wants to use dates outside `NaiveDate` range there is a wrapper `value::Date(u32)` for sending in queries. For receiving the user can just use raw `CQLValue`.
* Added more tests

### Time
* Added range validation
* Changed `std::time::Duration` to `chrono::Duration` to stay consistent with `Timestamp`
* For sending in queries there is a wrapper `value::Time(Duration)`
* Added more tests
* Maybe found a harmless scylla bug? `'-00:00:01'` is parsed as the time `0h 0m 1s`, while trying to insert `'-01:00:00'` fails

### Timestamp
* Changed representation to `chrono::Duration` - it was easy to confuse whether the timestamp was in seconds, milliseconds or nanoseconds. I'm not sure if this was a good decision. Ideally we could probably use something like [`chrono::DateTime`](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html) but I'm worried about compatibility:
`DateTime` is represented with:
```rust
date: NaiveDate, time: NaiveTime,
 ```
And our timestamp is represented in milliseconds since epoch.
Conversion between these types seems risky.
* Added more tests
* Scylla doesn't accept timestamps before `1970-01-01`, but the [specification](https://github.com/apache/cassandra/blob/78b13cd0e7a33d45c2081bb135e860bbaca7cbe5/doc/native_protocol_v4.spec#L929) allows it, is this intended?
* I tried using example timestamp values from cassandra [documentation](https://cassandra.apache.org/doc/latest/cql/types.html#timestamps), but scylla gives a different timestamp value, I guess it's just incompatible for some reason.

### Timeuuid
* Added tests

**Comment by dtzxporter** [comment]
Thoughts about pulling in `chrono` instead ? Its types match more consistently with cassandra:

Example: 
```
    Date(Date<Utc>),
    Time(Duration),
    Timestamp(DateTime<Utc>),
```

**Comment by cvybhu** [comment]
> Thoughts about pulling in `chrono` instead ? Its types match more consistently with cassandra:
> 
> Example:
> 
> ```
>     Date(Date<Utc>),
>     Time(Duration),
>     Timestamp(DateTime<Utc>),
> ```

This PR uses types from `chrono` already. 

For `Date` internal representation is `u32` because `chrono::NaiveDate` doesn't support full range of dates represented by `u32`.
Users can still send and receive dates as `chrono::NaiveDate` (see `examples/cql-time-types.rs`)

`Time` is represented as `chrono::Duration`

`Timestamp` is represented as `chrono::Duration` because `chrono::DateTime` internally keeps timestamp as a date and time and I'm worried about exact conversion between `milliseconds since epoch` and `date+time` with all these edge cases around leap seconds etc.

**Review by psarna — APPROVED** [review]

**Pull request merged**